connector/connect/common/src/main/protobuf/spark/connect/relations.proto

```diff
@@ -225,6 +225,16 @@ message Join {
     JOIN_TYPE_LEFT_SEMI = 6;
     JOIN_TYPE_CROSS = 7;
   }
+
+  // (Optional) Only used by joinWith. Set the left and right join data types.
+  optional JoinDataType join_data_type = 6;
+
+  message JoinDataType {
+    // If the left data type is a struct that can be flatten to a row.
+    bool is_left_flattenable_to_row = 1;
```

Comment:
i'd just name this `is_left_struct` or `is_left_single_value`. Flattening is a detail we don't really care about.

The same applies to the right side.